### PR TITLE
cmc score 2.0

### PR DESCRIPTION
**Pull request by julia-shenshina**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/callbacks/metric.py [hunk 1]
```diff
@@ -1,66 +1,132 @@
...
+        }
+
+    @staticmethod
+    def _validate_metric(metric: IMetric) -> None:
+        assert isinstance(metric, IMetric)
+
+    @staticmethod
+    def _format_keys_to_kv(keys: Union[str, Iterable[str], Dict[str, str]]) -> Dict[str, str]:
+        kv_keys = {}
+        if isinstance(keys, dict):
```
Comment (Scitator) [inline catalyst/callbacks/metric.py:116]: could u please check what happens in the `OrderedDict` case?

File: catalyst/callbacks/metric.py [hunk 2]
```diff
@@ -1,66 +1,132 @@
...
     def __init__(
-        self,
-        metric: ICallbackBatchMetric,
-        input_key: str,
-        target_key: str,
-        log_on_batch: bool = True,
+            self,
+            metric: Union[ICallbackBatchMetric, ICallbackLoaderMetric],
+            input_key: Union[str, Iterable[str], Dict[str, str]],
+            target_key: str,
```
Comment (Scitator) [inline catalyst/callbacks/metric.py]: is it possible to support KV for `target_key` too? :)

File: catalyst/callbacks/metric.py [hunk 3]
```diff
@@ -1,66 +1,132 @@
...
+        kv_inputs = {}
+        for key in self._keys:
+            kv_inputs[self._keys[key]] = runner.engine.sync_tensor(runner.batch[key])
+        return kv_inputs
+
+    def on_batch_end(self, runner: "IRunner") -> None:
+        raise NotImplementedError
+
+    def on_loader_start(self, runner: "IRunner") -> None:
+        raise NotImplementedError
```
Comment (Scitator) [inline catalyst/callbacks/metric.py]: should we make them abstract?

Comment (Scitator) [reply catalyst/callbacks/metric.py]: btw, could u please check the order of the methods:
- on_loader_start
- on_batch_start
- on_batch_end
- on_loader_end

File: catalyst/callbacks/metric.py (hunk 2 above)
Comment (Scitator) [reply catalyst/callbacks/metric.py]: and work in 2 cases only:
- str for the input and str for the output
- key-value for the input and key-value for the output
?
what do u think about it?

Comment (Scitator) [reply catalyst/callbacks/metric.py]: is such case it would be easier to redefine the `on_batch_end` logic during the `__init__` and reduce the number of if statements during the run

File: catalyst/metrics/_cmc_score.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        self.accumulative_fields = accumulative_fields
+        self.storage = None
+        self.num_samples = None
+        self.collected_batches = None
+        self.tmp_idx = None
+
+    def reset(self, num_batches: int, num_samples: int) -> None:
+        self.num_samples = num_samples
+        self.collected_batches = 0
+        self.tmp_idx = 0
```
Comment (Scitator) [inline catalyst/metrics/_cmc_score.py]: `collected_samples`?

File: catalyst/metrics/_cmc_score.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        self.labels_key = labels_key
+        self.is_query_key = is_query_key
+        self.topk_args = topk_args
+        self.metric_name = "cmc"
+
+    def compute(self) -> List[float]:
+        query_embeddings = self.storage[
+            self.embeddings_key
+        ][self.storage[self.is_query_key] == 1].float()
+        query_labels = self.storage[self.labels_key][self.storage[self.is_query_key] == 1]
```
Comment (Scitator) [inline catalyst/metrics/_cmc_score.py]: could we do something like `query_mask = self.storage[self.is_query_key] == 1` and reuse it?

Comment (Scitator) [reply catalyst/metrics/_cmc_score.py]: the same for the `== 0` case

File: catalyst/metrics/_cmc_score.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                topk=k,
+            )
+            metrics.append(value)
+
+        return metrics
+
+    def compute_key_value(self) -> Dict[str, float]:
+        values = self.compute()
+        kv_metrics = {
+            f"{self.metric_name}{k:02d}": value for k, value in zip(self.topk_args, values)
```
Comment (Scitator) [inline catalyst/metrics/_cmc_score.py:265]: could u please add `prefix` and `suffix` here?
they are required in case when we have a multi-head network and want to compute the metrics for each of them.

File: catalyst/callbacks/metric.py (hunk 1 above)
Comment (julia-shenshina) [reply catalyst/callbacks/metric.py:116]: isinstance(OrderedDict(), dict) is True

Comment (Scitator) [reply catalyst/callbacks/metric.py:116]: 🤔  what about python 3.6?)

File: catalyst/callbacks/metric.py [hunk 7]
```diff
@@ -1,78 +1,300 @@
...
 from catalyst.core.callback import Callback, CallbackNode, CallbackOrder
 from catalyst.core.runner import IRunner
-from catalyst.metrics._metric import ICallbackBatchMetric, ICallbackLoaderMetric
+from catalyst.metrics import (
+    ICallbackBatchMetric,
+    ICallbackLoaderMetric,
+    IMetric,
+)
+
+TORCH_BOOL = torch.bool if torch.__version__ > "1.1.0" else torch.ByteTensor
```
Comment (Scitator) [inline catalyst/callbacks/metric.py]: let's just use `TORCH_BOOL`?)

File: catalyst/callbacks/metric.py [hunk 8]
```diff
@@ -1,78 +1,300 @@
...
+            runner.batch[self.input_key],
+            runner.batch[self.target_key],
+        )
+        inputs, targets = (
+            runner.engine.sync_tensor(inputs),
+            runner.engine.sync_tensor(targets),
+        )
+        return inputs, targets
+
+    def _get_key_value_inputs(
```
Comment (Scitator) [inline catalyst/callbacks/metric.py]: btw, the good thing about docs – you don't need to write them for the private methods.

File: catalyst/callbacks/metric.py [hunk 9]
```diff
@@ -1,78 +1,300 @@
...
+        Args:
+            metric: metric to calculate in callback
+            input_key: keys of tensors that should be used as inputs in metric calculation
+            target_key: keys of tensors that should be used as targets in metric calculation
+        """
         super().__init__(order=CallbackOrder.metric, node=CallbackNode.all)
-        assert isinstance(metric, ICallbackBatchMetric)
         self.metric = metric
+        self._validate_metric(metric=metric)
+        self._metric_update_method = self.metric.update
```
Comment (Scitator) [inline catalyst/callbacks/metric.py:68]: why do we need it?

Comment (Scitator) [reply catalyst/callbacks/metric.py:68]: I see, that u are using `self.metric.update` and `self.metric.update_key_value`, but it would be great if you could explain the case also :)

File: catalyst/metrics/__init__.py [hunk 10]
```diff
@@ -21,7 +21,7 @@
     MulticlassPrecisionRecallF1SupportMetric,
     MultilabelPrecisionRecallF1SupportMetric,
 )
-
+from catalyst.metrics._cmc_score import CMCMetric
```
Comment (Scitator) [inline catalyst/metrics/__init__.py]: should we also add `AccumulationMetric` here?

File: catalyst/metrics/_cmc_score.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+    def __init__(
+        self,
+        compute_on_call: bool = True,
+        prefix: Optional[str] = None,
+        suffix: Optional[str] = None,
+        embeddings_key: str = "features",
+        labels_key: str = "targets",
+        is_query_key: str = "is_query",
+        topk_args: Iterable[int] = (1,),
```
Comment (Scitator) [inline catalyst/metrics/_cmc_score.py]: could we make it `None` and do `self.topk_args = topk_args or (1,)` later?

Comment (Scitator) [inline catalyst/metrics/_cmc_score.py]: btw, could we move them on top of compute_on_call/prefix/suffix as far as they are more important for this metric?

File: catalyst/metrics/tests/test_cmc_score.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+import numpy as np
+import pytest
+
+import torch
+
+from catalyst.metrics._cmc_score import AccumulationMetric, CMCMetric
+
+
+@pytest.fixture()
```
Comment (Scitator) [inline catalyst/metrics/tests/test_cmc_score.py:14]: just try
```suggestion
@pytest.fixture()  # noqa: WPS320 
```

File: catalyst/metrics/tests/test_cmc_score.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                    )
+                batch_data[field_name] = data
+            batches.append(batch_data)
+        batched_data.append(
+            (fields_names, num_batches, num_samples, batches, true_values)
+        )
+    return batched_data
+
+
+def test_accumulation(generate_batched_data) -> None:
```
Comment (Scitator) [inline catalyst/metrics/tests/test_cmc_score.py]:
```suggestion
def test_accumulation(generate_batched_data) -> None:  # noqa: WPS442 
```
